dnf5daemon-server/services/rpm/rpm.cpp

```diff
@@ -392,6 +408,64 @@ sdbus::MethodReply Rpm::list(sdbus::MethodCall & call) {
     return reply;
 }
 
+void Rpm::list_fd(sdbus::MethodCall & call) {
+    // read options from dbus call
+    dnfdaemon::KeyValueMap options;
+    call >> options;
+    // get the output pipe file descriptor
+    sdbus::UnixFd dbus_unix_fd;
+    call >> dbus_unix_fd;
+    // id of the output filedescriptor
+    std::string out_fd_id;
+    call >> out_fd_id;
+
+    int out_fd = dbus_unix_fd.get();
+
+    session.fill_sack();
+
+    auto query = filter_packages(options);
+
+    std::vector<std::string> package_attrs =
+        dnfdaemon::key_value_map_get<std::vector<std::string>>(options, "package_attrs", std::vector<std::string>());
+    std::string error_msg{};
+    for (const auto & pkg : query) {
+        std::string jsonstr;
+        try {
+            jsonstr = package_to_json(pkg, package_attrs) + "\n";
+        } catch (const std::exception & ex) {
+            error_msg = fmt::format(
+                "Error serializing package \"{0}\" from repo \"{1}\": {2}",
+                pkg.get_nevra(),
+                pkg.get_repo_id(),
+                ex.what());
+        } catch (...) {
+            error_msg = fmt::format(
+                "Unknown exception caught while serializing package \"{0}\" from repo \"{1}\"",
+                pkg.get_nevra(),
+                pkg.get_repo_id());
+        }
+        if (error_msg.empty()) {
+            std::string write_error;
+            if (!dnfdaemon::write_to_fd(jsonstr, out_fd, write_error)) {
+                error_msg = fmt::format("Error writing package list to the fd: {}", write_error);
+                break;
+            }
+        }
+    }
+    close(out_fd);
+
+    // signal client that the transfer has finished and the output file descriptor is closed
+    auto dbus_object = get_session().get_dbus_object();
+    auto signal = dbus_object->createSignal(call.getInterfaceName(), dnfdaemon::SIGNAL_WRITE_TO_FD_FINISHED);
+    signal << error_msg.empty();
+    signal << out_fd_id;
+    signal << error_msg;
+    try {
+        dbus_object->emitSignal(signal);
+    } catch (...) {
```

Comment:
Shell we log something on the server side?

Reply:
Yes. Let me just add a TODO for now, because there are more places which need logging (already marked with TODO). I would like to fix them in follow up PR.